### ISSUE TITLE
Create a CMake project file

### ISSUE THREAD

**Issue by cmumford** [lvgl/lv_lib_split_jpg#24]
Create a CMake project to make it simple to incorporate lv_lib_split_jpg into other project. It also needs to support a standalone build for continuous integration and desktop development.

### When included by a higher level CMake file:
1. On ESP-IDF: define an ESP-IDF component.
2. On desktop OS: build a static library.
3. Use the LVGL provided by the higher level project.
4. Use the LVGL settings (`lv_conf.h`, `LV_LVGL_H_INCLUDE_SIMPLE`, etc.) defined by the higher level project.

### When used as a top level CMake file:
1. Build for:
   1. Linux.
   2. Windows.
   3. macOS.
   4. ESP-IDF (stretch goal).
2. Build against specified LVGL:
   1. lvgl/master branch for lv_lib_split_jpg/master
   2. lvgl/{branch} for lv_lib_split_jpg/{branch}
3. Build example program(s).
4. Build test program(s).

**Comment by kisvegabor** [comment]
Hi Chris,

Sounds good!

**Some questions**

>Use the LVGL provided by the higher level project.

How do you mean it?

> When used as a top level CMake file:
Build for: ...

The result could be a `DLL`, `so`, etc file made available as build artifact of the CI? (With a link in the README?)

> Build test program(s)

We can reuse the [Unity test framework](https://github.com/lvgl/lvgl/tree/master/tests) developed for the core LVGL.

---

We should also find a  suitable release and versioning model for these external library projects. Now there is no versioning here, just randomly updated to work with the latest LVGL. Things to consider:
- Follow major versions of LVGL?
- Follow minor version LVGL too?
- Have release branches for minor versions?
- Merge back fixes to the minor release branches?
- What can to automate from these?

**Comment by cmumford** [comment]
Hi Gabor:

So I'm making good progress. I've got the workflows defined and working. Here's a run where they [all passed](https://github.com/cmumford/lv_lib_split_jpg/runs/3297104746). I then injected a [memory bug](https://github.com/cmumford/lv_lib_split_jpg/commit/d5109e4ca6ed71c66726a62397524fcb9e2abd6e) to verify the ASAN tests [failed](https://github.com/cmumford/lv_lib_split_jpg/runs/3297117640) as expected.

> > Use the LVGL provided by the higher level project.
> 
> How do you mean it?

At present, **on desktop**, if you checked out my dev branch and ran cmake it will pull **it's own copy of LVGL** (master branch) and build against it. This creates a static library (`liblv_lib_split_jpg.a`) that expects to be linked with `liblvgl.a` build from master. I think we need to be able to support a project which wants to build lv_lib_split_jpg and, when it compiles, includes the outer projects `lvgl.h`, etc.

For example we (hypothetically) we want to have lv_sim_eclipse_sdl build and use lv_lib_split_jpg. lv_sim_eclipse_sdl gets a copy of lvgl and lv_lib_split_jpg (via a git submodule), and when it builds lv_lib_split_jpg we don't want that cloning a second copy of lvgl. It needs to compile against lv_sim_eclipse_sdl's lvgl submodule.

> > When used as a top level CMake file:
> > Build for: ...
> 
> The result could be a `DLL`, `so`, etc file made available as build artifact of the CI? (With a link in the README?)

Yes, right now they are static libraries, but could be shared, or both. We would use the release mechanism provided by GitHub: https://github.com/lvgl/lv_lib_split_jpg/releases.

> 
> > Build test program(s)
> 
> We can reuse the [Unity test framework](https://github.com/lvgl/lvgl/tree/master/tests) developed for the core LVGL.

Absolutely. They could be authored any way we wish. As long as they return a non-zero return code upon failure they will work for GitHub's CI.

> We should also find a suitable release and versioning model for these external library projects. Now there is no versioning here, just randomly updated to work with the latest LVGL. Things to consider:
> 
> * Follow major versions of LVGL?
> * Follow minor version LVGL too?
> * Have release branches for minor versions?
> * Merge back fixes to the minor release branches?
> * What can to automate from these?

I personally find it difficult to deal with the number of repos when doing LVGL development, and anything we can do to simplify this will be helpful (at least to me). I would recommend following LVGL's branch names at a minimum. I would even go so far as to branch every LVGL project on a major release, even if there were no changes in one of the satellite projects (e.g. lv_fs_if, etc.).

I think there's a fair amount of automation that could be done, but this can be added after the build/test workflows are working. One thing I would like to see is a triggering of all workflows for downstream projects for a LVGL commit.

**Comment by embeddedt** [comment]
A TL;DR summary is at the bottom because I ended up writing a very long message. :wink: 

------------

There is quite a bit of complexity involved in keeping everything synchronized between many repositories.

Case in point: during the v8 development cycle, it was quite common for a commit to be pushed to the main lvgl repository which changed an API, but the corresponding change was not made to lv_examples. This meant that `lv_sim_emscripten`, which pulls the head of both repositories every two hours, frequently failed as the two APIs were not synchronized. Having a monorepo discourages this type of issue because it is possible to make an atomic commit where the API and examples are changed at once.

Had we set up unit tests running across repositories (e.g. on `lv_examples`), they would also have run into this issue.

Another example: the docs used to be in a separate repository. Builds used to be done by @kisvegabor manually, but this resulted in users waiting a long time for rebuilds, so eventually I decided to automate the process with a GitHub action.

The automated builds would pull `lv_examples` so that inline examples could be generated in the HTML. Once again, however, the logic to do this was relatively complex, as there were at least three repositories involved in the overall process:

* `lv_examples` main branch, for the source code.
* `lv_examples` `gh-pages` branch, for the built HTML files.
* `lv_sim_emscripten` - used by `lv_examples` to build the HTML files.

Any incompatibility between these would result in build failure, which typically required me to manually fix scripts/bump a submodule somewhere.

Contribution to the docs was also not the smoothest process. When a contribution was submitted to lvgl/lvgl that involved a new feature/API tweak, we usually blocked it and waited for a PR to be opened to lvgl/docs as well. This was to ensure that the docs would actually get updated at the same time that the change was brought in. 

Now that we merged the docs into the main repository with v8, this is not an issue, as a contributor can update the docs at the same time.

**TL;DR** Automated repository synchronization is not trivial to do correctly. What I did for v7 wasn't the cleanest implementation, and I much prefer the new repository model in v8, as it greatly simplified the various workflows, so I'm wondering whether it's worth pursuing this for the remaining "satellite projects". Atomicity between project components greatly reduces the opportunities for failure.

**Comment by kisvegabor** [comment]
So we have already merged `docs` (source md files) and `examples` into `lvgl` which really made things much simpler.  I started to think about what if we merge `lv_demos` and 3rd party library projects (like this jpeg repo) into `lvgl`. The project structure can be kept clear. Although the growing repo size could be an issue. 

People started to work on a new driver layer and ready to use display and indev drivers for LVGL. Probably the drivers could be merged too.

I've hear that Google is big advocate of monorepos. @cmumford, your profile says are working at Google. Can you share your experiences about it?  (Of course, if it doesn't conflicts with your NDA).

**Comment by cmumford** [comment]
I would be strongly in favor of getting as close to a single repository as possible. I think that separate repos make sense when they can be used for a completely non-LVGL project. I work on Chrome, and several years back we merged Blink (a fork of WebKit) into Chromium. Prior to the merge changing an API which was used in the other repo required a minimum of three commits. Say, for example we wanted to change something like `int sum(int a, int b)` to `float sum(float a, float b)`, it would be done in three commits:

1. Land `float sum(float a, float b)` as a second function **beside** `int sum(int a, int b)` in the first commit, and change all references in the repo in which that is defined.
2. Switch to `float sum(float a, float b)` in the second repo.
3. Delete `int sum(int a, int b)` in the first repo.

Unfortunately we can't override functions in C, so there is no way to do this without changing function names, or temporarily breaking one of the two repos. If the goal is to have PRs uploaded, automatically tested, and merged w/o breaking anything, then it all needs to be done in a single repo.

**Comment by kisvegabor** [comment]
> I work on Chrome

I hoped that you are using LVGL at Google :slightly_smiling_face: 

---

As I'm thinking about it more the idea of the mono repo makes more and more sense. I've opened [an issue](https://github.com/lvgl/lvgl/issues/2456) it to get more thought about it.

**Comment by cmumford** [comment]
Sadly I am not using LVGL at Google - only personal projects. Thx for opening the mono repo issue. Looking forward to reviewing it. As https://github.com/lvgl/lv_lib_split_jpg/pull/27 has merged I will close this issue.
